source/shared/core_conn.cpp

```diff
@@ -150,34 +150,125 @@ sqlsrv_conn* core_sqlsrv_connect( _In_ sqlsrv_context& henv_cp, _In_ sqlsrv_cont
+            // https://docs.microsoft.com/en-us/sql/connect/odbc/linux-mac/installing-the-microsoft-odbc-driver-for-sql-server#microsoft-odbc-driver-11-for-sql-server-on-linux
+
+            DRIVER_VERSION odbc_version = ODBC_DRIVER_UNKNOWN;
+            if( core_search_odbc_driver_unix( ODBC_DRIVER_13 ) ) {
+                odbc_version = ODBC_DRIVER_13;
+            } 
+            else if ( core_search_odbc_driver_unix( ODBC_DRIVER_17 ) ) {
+                odbc_version = ODBC_DRIVER_17;
+            }
+
+            CHECK_CUSTOM_ERROR( odbc_version == ODBC_DRIVER_UNKNOWN, conn, SQLSRV_ERROR_DRIVER_NOT_INSTALLED, get_processor_arch() ) {
+                throw core::CoreException();
+            }
+            std::string conn_str_driver = conn_str + CONNECTION_STRING_DRIVER_NAME[odbc_version];
+            r = core_odbc_connect( conn, conn_str_driver, is_pooled );
+        }  // else driver_version not unknown
+    } // else ce_option enabled
+#else
+            r = core_odbc_connect( conn, conn_str, is_pooled );
+
+            if( core_compare_error_state( conn, r, "IM002" ) ) {
+                // sql state IM002 means that the specified ODBC driver is not installed
+                CHECK_CUSTOM_ERROR( true, conn, SQLSRV_ERROR_SPECIFIED_DRIVER_NOT_FOUND ) {
+                    throw core::CoreException();
+                }
+            }
+        }
+        else {
+            // driver not specified, so connect using ODBC 17
+            conn_str = conn_str + CONNECTION_STRING_DRIVER_NAME[ODBC_DRIVER_17];
+            r = core_odbc_connect( conn, conn_str, is_pooled );
 
-        for ( std::size_t i = DRIVER_VERSION::FIRST; i <= DRIVER_VERSION::LAST; ++i ) {
-            is_missing_driver = false;
-            std::string conn_str_driver = conn_str + CONNECTION_STRING_DRIVER_NAME[ DRIVER_VERSION(i) ];
-            r = core_odbc_connect( conn, conn_str_driver, is_missing_driver, is_pooled );
-            CHECK_CUSTOM_ERROR( is_missing_driver && ( i == DRIVER_VERSION::LAST ), conn, SQLSRV_ERROR_DRIVER_NOT_INSTALLED, get_processor_arch()) {
+            // sql state IM002 means that the specified ODBC driver is not installed
+            CHECK_CUSTOM_ERROR( core_compare_error_state( conn, r, "IM002" ) , conn, SQLSRV_ERROR_CE_DRIVER_REQUIRED, get_processor_arch() ) {
                 throw core::CoreException();
             }
-            if ( !is_missing_driver) {
-                break;
+        } 
+    }
+    else { 
+        // column encryption NOT enabled
+        if( conn->driver_version != ODBC_DRIVER_UNKNOWN ) {
+            r = core_odbc_connect( conn, conn_str, is_pooled );
+
+            if( core_compare_error_state( conn, r, "IM002" ) ) {
+                // sql state IM002 means that the specified ODBC driver is not installed
+                CHECK_CUSTOM_ERROR( true, conn, SQLSRV_ERROR_SPECIFIED_DRIVER_NOT_FOUND, CONNECTION_STRING_DRIVER_NAME[conn->driver_version].c_str() ) {
```

Comment:
1. please be more consistent -
  sometimes we do core_compare_error_state() then CHECK_CUSTOM_ERROR(true, blah)
  sometimes we do CHECK_CUSTOM_ERROR(core_compare_error_state(), blah)
  the latter is preferable IMHO
2. there are multiple places where this is done, can we refactor this so it's not in multiple places??